### DISCUSSION TITLE
How to make my library tree-shakable?

### DISCUSSION THREAD

**Discussion by lucypoly** [webpack/webpack#14808] [Q&A]
Hello, I'm using webpack 5 and typescript for my library. How to make it tree-shakable and build with esm? Can't see any info in docs. I'm using:
```
"webpack": "5.64.2",
"webpack-cli": "4.9.1"
 ```

**Comment by alexander-akait** [comment]
You don't need webpack to build ESM library, just setup typescript to use ESM

**Reply by lucypoly** [reply]
But with webpack 4 it wasn't possible #2933 (comment)
Now is it possible?

**Reply by alexander-akait** [reply]
Why do you need webpack to build library? Typescript can output ESM code, so you don't need extra tool to create ESM library, just setup typescript to output ESM, they have docs https://www.typescriptlang.org/docs/handbook/esm-node.html

**Reply by lucypoly** [reply]
Because it is built with webpack a long time ago, and I shouldn't change that (we need it to build CSS, files and so on). I only need to make it tree-shakable.

**Reply by alexander-akait** [reply]
Double bundling, i.e. bundle after another bundler has limitations, tree-shakable is not possible

**Reply by lucypoly** [reply]
What do you mean by double boundle? Why is it double? 
So I want to use the new feature for webpack "type": 'module', I can't? 
I want to have the feature that is described here https://github.com/webpack/webpack/issues/2933

**Reply by alexander-akait** [reply]
> What do you mean by double boundle? Why is it double?

Most of applications use bundler, if you bundle your library, other developers which will use your library also will bundle it inside your application - it is double bundling, you should avoid it because it will kill many optimizations

**Reply by lucypoly** [reply]
So do you mean I can't use your new feature `type: 'module`? It won't help?

**Reply by alexander-akait** [reply]
Firstly - why you need webpack to bundle library?

**Reply by lucypoly** [reply]
Because it is already set up with webpack and I don't want to do all the staff from the very beginning. It has a great number of configurations, and also I'm not the owner of the lib, I only have a task to make it tree-shakable

**Reply by alexander-akait** [reply]
It is impossible right now and you really don't need webpack to do ESM library

**Reply by lucypoly** [reply]
Strange that there are so many articles how to do tree-shakable library with webpack https://blog.theodo.com/2021/04/library-tree-shaking/

**Reply by alexander-akait** [reply]
Article tells how to do library ready for tree-shakable which will be bundled by webpack/rollup, it is not about hot to create library using webpack

**Reply by lucypoly** [reply]
No, he creates his own `"user-library"` and making it tree-shakable

**Reply by lucypoly** [reply]
@alexander-akait  Maybe you know why there could be an error in my final app
```
./node_modules/@amblyosight/lib-components/dist/index.js 53696:31
Module parse failed: Unexpected token (53696:31)
File was processed with these loaders:
* ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
```
after I added `module: true` to webpack config of my library?

**Reply by alexander-akait** [reply]
Sorry, don't know, you really don't need webpack to build standalone ESM library, just setup typescript to emit ESM
